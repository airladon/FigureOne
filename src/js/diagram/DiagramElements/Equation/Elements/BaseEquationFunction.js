
// @flow
import {
  Point,
} from '../../../../tools/g2';
import {
  DiagramElementPrimitive, DiagramElementCollection,
} from '../../../Element';
import { duplicateFromTo } from '../../../../tools/tools';
import { Element, Elements } from './Element';
// import Bounds from './Bounds';

export default class BaseEquationFunction extends Elements {
  contents: Array<Elements | null>;
  glyphs: Array<DiagramElementPrimitive | DiagramElementCollection | null>;
  glyphLocations: Array<Point>;
  glyphWidths: Array<number>;
  glyphHeights: Array<number>;
  // inSize: boolean;
  options: Object;

  constructor(
    content: Elements | null | Array<Elements | null>,
    glyph: DiagramElementPrimitive | null | DiagramElementCollection
      | Array<DiagramElementPrimitive | null | DiagramElementCollection>,
    // inSize: boolean = true,
    options: Object,
  ) {
    const glyphElements = [];
    if (Array.isArray(glyph)) {
      glyph.forEach((g) => {
        glyphElements.push(g !== null ? new Element(g) : null);
      });
    } else {
      glyphElements.push(glyph !== null ? new Element(glyph) : null);
    }
    const glyphs = [];
    if (Array.isArray(glyph)) {
      glyphs = glyph;
    } else {
      glyphs.push(glyph);
    }

    let contentArray: Array<Elements | null> = [];
    if (Array.isArray(content)) {
      contentArray = content;
    } else {
      contentArray.push(content);
    }
    super([...glyphElements, ...contentArray]);
    this.glyphs = glyphs;
    this.contents = contentArray;
    this.glyphLocations = glyphElements.map(() => new Point(0, 0));
    this.glyphWidths = glyphElements.map(() => 1);
    this.glyphHeights = glyphElements.map(() => 1);
    this.options = options;
    // this.inSize = inSize;
  }

  _dup(namedCollection?: Object) {
    const copyContent = this.contents.map(
      content => (content == null ? null : content._dup(namedCollection)),
    );
    let { glyphs } = this;
    // let copyGlyphs = [];
    if (namedCollection) {
      const newGlyphs = [];
      this.glyphs.forEach((g) => {
        if (g != null) {
          newGlyphs.push(namedCollection[g.name]);
        } else {
          newGlyphs.push(g);
        }
      });
      glyphs = newGlyphs;
    }
    const copy = new this.constructor(
      copyContent,
      glyphs,
      // this.inSize,
      this.options,
    );
    duplicateFromTo(
      this, copy,
      ['content', 'contents', 'glyphs'],
    );
    return copy;
  }

  getAllElements() {
    let elements = [];
    this.contents.forEach((c) => {
      if (c != null) {
        elements = [...elements, ...c.getAllElements()];
      }
    });
    this.glyphs.forEach((g) => {
      if (g != null) {
        elements = [...elements, g];
      }
    });
    return elements;
  }

  setPositions() {
    this.glyphs.forEach((glyph, index) => {
      if (glyph != null) {
        const t = glyph.getTransform()._dup();
        t.updateTranslation(this.glyphLocations[index].x, this.glyphLocations[index].y);
        t.updateScale(this.glyphWidths[index], this.glyphHeights[index]);
        glyph.setTransform(t);
      }
    });
    this.contents.forEach((content) => {
      if (content != null) {
        content.setPositions();
      }
    });
  }

  offsetLocation(offset: Point = new Point(0, 0)) {
    this.location = this.location.add(offset);
    this.glyphLocations.forEach((glyphLocation, index) => {
      if (this.glyphs[index] != null) {
        this.glyphLocations[index] = glyphLocation.add(offset);
      }
    });
    this.contents.forEach((content) => {
      if (content != null) {
        content.offsetLocation(offset);
      }
    });
  }

  // eslint-disable-next-line class-methods-use-this, no-unused-vars
  calcSize(location: Point, scale: number) {
  }
}
