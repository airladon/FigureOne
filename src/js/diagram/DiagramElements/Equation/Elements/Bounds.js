// @flow
export default class Bounds {
  width: number;
  height: number;
  ascent: number;
  descent: number;
  constructor() {
    this.width = 0;
    this.height = 0;
    this.ascent = 0;
    this.descent = 0;
  }
}
