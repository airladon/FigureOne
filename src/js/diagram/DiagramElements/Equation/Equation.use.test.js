import {
  Point,
} from '../../../tools/g2';
import {
  round,
} from '../../../tools/math';
import * as tools from '../../../tools/tools';
// import * as colorTools from '../../../tools/color';
import makeDiagram from '../../../__mocks__/makeDiagram';
import { Equation } from './Equation';
// import EquationForm from './EquationForm';
import { Elements } from './Elements/Element';
import * as html from '../../../tools/htmlGenerator';
// import Fraction from './Elements/Fraction';

tools.isTouchDevice = jest.fn();

jest.mock('../../Gesture');
jest.mock('../../webgl/webgl');
jest.mock('../../DrawContext2D');

// const cleanUIDs = (objectToClean: {}) => {
//   const genericUID = '0000000000';
//   if (objectToClean == null) {
//     return;
//   }
//   if (objectToClean.uid != null) {
//     if (objectToClean.uid === genericUID) {
//       return;
//     }
//     objectToClean.uid = genericUID;
//   }
//   const keys = Object.keys(objectToClean);
//   for (let i = 0; i < keys.length; i += 1) {
//     const key = keys[i];
//     const value = objectToClean[key];
//     if (
//       typeof value === 'object'
//       && !Array.isArray(value)
//       && value != null
//       && typeof value !== 'function'
//       && typeof value !== 'number'
//       && typeof value !== 'boolean'
//       && typeof value !== 'string'
//       ) {
//         cleanUIDs(value);
//     }
//   };
// }

/* eslint-disable no-param-reassign */
const cleanElements = (elements) => {
  const r = 8;
  elements.ascent = round(elements.ascent, r);
  elements.descent = round(elements.descent, r);
  elements.width = round(elements.width, r);
  elements.height = round(elements.height, r);
  if (elements.location) {
    elements.location = elements.location.round(r);
  }
  if (Array.isArray(elements.content)) {
    elements.content.forEach((c) => {
      cleanElements(c);
    });
    // elements.content = elements.content.map(c => cleanElements(c));
  } else if (
    elements.content instanceof Elements
    || elements.content instanceof Element
  ) {
    cleanElements(elements.content);
  } else {
    elements.content = [];
  }
};

function cleanForm(form) {
  form.collectionMethods = {};
  form.elements = {};
  form.name = 'name';
  cleanElements(form);
}
/* eslint-enable no-param-reassign */

describe('Different ways to make an equation', () => {
  let diagram;
  let eqn;
  let color1;
  let color2;
  let ways;
  let clean;
  beforeEach(() => {
    clean = (formName) => {
      cleanForm(eqn.eqn.forms[formName]);
      return eqn.eqn.forms[formName];
    };
    diagram = makeDiagram();
    color1 = [0.95, 0, 0, 1];
    color2 = [0, 0.95, 0, 1];
    ways = {
      allTextInConstructor: () => {
        eqn = new Equation(diagram.shapes, {
          color: color1,
          elements: {
            a: 'a',
            b: 'b',
            c: 'c',
            _2: '2',
            v: { symbol: 'vinculum' },
          },
          forms: {
            0: ['a', 'b', 'c'],
            1: [{ frac: ['a', 'v', '_2'] }, 'c'],
          },
          formSeries: ['0', '1'],
        });
      },
      allTextInConstructorAllOptionsNew__TO_TEST__TODO: () => {
        eqn = new Equation(diagram.shapes, {
          color: color1,
          position: [1, 1],           // Points can be defined as arrays
          elements: {
            a: 'a',
            b: { color: [1, 1, 0, 1] }, // Element text same as key name
            c: 'c',
            _2: '2',
            v: { symbol: 'vinculum' },
          },
          scale: 0.45,
          formDefaults: {
            alignment: {
              fixTo: { x: 2, y: 2 },    // Points can also be defined as objects
              xAlign: 'right',
              yAlign: 'top',
            },
            animation: {
              duration: 1,
              translation: {
                c: { style: 'curved', direction: 'up', mag: 0.5 },
              },
            },
            elementMods: {
              b: { color: [1, 0, 0, 1] },
            },
          },
          // Phrases are combinations of elements that can be used in forms to
          // make the forms simpler
          phrases: {
            ab: ['a', 'b'],
          },
          forms: {
            0: ['a', 'b', 'c'],
            1: [{ frac: ['a', 'v', '_2'] }, 'c'],
            // Equals can be used as an object key, so it is a valid inline definition
            2: ['ab', '=', 'c'],
            // Full form definition
            3: {
              content: ['c', 'b', 'd'],       // New elements can be defined (d)
              scale: 1.2,
              elementMods: {
                b: { color: [0, 1, 0, 1] },
              },
              description: '|Form| 2 |description|',
              modifiers: {
                Form: html.highlight([1, 0, 0, 0]),
              },
              animation: {
                duration: null,
                translation: {
                  c: { style: 'curved', direction: 'down', mag: 0.5 },
                },
              },
              alignment: {
                fixTo: 'b',
                xAlign: 'center',
                yAlign: 'baseline',
              },
              fromForm: {
                2: {
                  animation: {
                    duration: 2,
                    translation: {
                      c: { style: 'linear' },
                    },
                  },
                  elementMods: {
                    b: { color: [0, 0, 1, 1] },
                  },
                },
              },
            },
          },
          // Either a single or multiple form series can be defined
          formSeries: {
            a: ['0', '1'],
            b: ['1', '2'],
          },
        });
      },
      allTextInConstructorAllOptions: () => {
        eqn = new Equation(diagram.shapes, {
          color: color1,
          position: [1, 1],           // Points can be defined as arrays
          elements: {
            a: 'a',
            b: 'b',
            c: 'c',
            _2: '2',
            v: { symbol: 'vinculum' },
          },
          formDefaults: {
            alignment: {
              fixTo: { x: 2, y: 2 },    // Points can also be defined as objects
              xAlign: 'right',
              yAlign: 'top',
            },
          },
          scale: 0.45,
          forms: {
            0: ['a', 'b', 'c'],
            1: [{ frac: ['a', 'v', '_2'] }, 'c'],
          },
          formSeries: ['0', '1'],
        });
      },
      separateAllText: () => {
        eqn = new Equation(diagram.shapes, { color: color1 });
        eqn.addElements({
          a: 'a',
          b: 'b',
        });
        eqn.addElements({
          a: 'a',
          b: 'b',
          c: 'c',
          _2: '2',
          v: { symbol: 'vinculum' },
        });
        eqn.addForms({
          0: ['a', 'b', 'c'],
          1: [{ frac: ['a', 'v', '_2'] }, 'c'],
        });
        eqn.addForms({
          2: ['a', 'b', 'c'],
        });
        eqn.setFormSeries(['0', '1', '2']);
        eqn.showForm('1');
      },
      phrases: () => {
        eqn = new Equation(diagram.shapes, { color: color1 });
        const frac = eqn.eqn.functions.frac.bind(eqn.eqn.functions);
        eqn.addElements({
          a: 'a',
          b: 'b',
          c: 'c',
          _2: '2',
          v: { symbol: 'vinculum' },
        });
        eqn.addPhrases({
          // Simple
          abc: ['a', 'b', 'c'],
          // All different ways to write a fraction
          f0: { frac: ['a', 'v', 'b'] },
          f1: {
            frac: {
              numerator: 'a',
              denominator: 'b',
              symbol: 'v',
            },
          },
          f2: frac({
            numerator: 'a',
            denominator: 'b',
            symbol: 'v',
          }),
          f3: frac(['a', 'v', 'b']),
          // Phrases dependent on other phrases
          ba: ['b', 'a'],
          bac: ['ba', 'c'],
          //
          // Phrase for integration into fraction
          ab: ['a', 'b'],
        });
        eqn.addForms({
          '0a': ['a', 'b', 'c'],
          '0b': 'abc',
          //
          f: { frac: ['a', 'v', 'b'] },
          f0: 'f0',
          f1: 'f1',
          f2: 'f2',
          f3: 'f3',
          //
          d: ['b', 'a', 'c'],
          d0: 'bac',
          //
          e: { frac: [['a', 'b'], 'v', 'c'] },
          e0: { frac: ['ab', 'v', 'c'] },
        });
      },
      equationScale: () => {
        eqn = new Equation(diagram.shapes, {
          color: color1,
          scale: 0.95,
        });
        eqn.addElements({
          a: 'a',
          b: 'b',
          c: 'c',
        });
        eqn.addForms({
          0: ['a', 'b'],
          1: {
            content: ['a', 'b'],
            scale: 0.85,
          },
        });
      },
      equationFormAlignment: () => {
        eqn = new Equation(diagram.shapes, {
          color: color1,
          formDefaults: {
            alignment: {
              fixTo: 'a',
              yAlign: 'top',
              xAlign: 'center',
            },
          },
        });
        eqn.addElements({
          a: 'a',
          b: 'b',
          c: 'c',
        });
        eqn.addForms({
          aCenterTop: ['a', 'space1', 'b'],
          originBasline: {
            content: ['a', 'space1', 'b'],
            alignment: {
              fixTo: new Point(0, 0),
              xAlign: 'left',
              yAlign: 'baseline',
            },
          },
          aRightMiddle: {
            content: ['a', 'space1', 'b'],
            alignment: {
              fixTo: 'a',
              xAlign: 'right',
              yAlign: 'middle',
            },
          },
          bRightBottom: {
            content: ['a', 'space1', 'b'],
            alignment: {
              fixTo: 'b',
              xAlign: 'right',
              yAlign: 'bottom',
            },
          },
        });
      },
      autoAddElements: () => {
        eqn = new Equation(diagram.shapes, {
          color: color1,
          scale: 0.95,
        });
        eqn.addForms({
          0: [
            'a',
            'b_1',
            '_c_2',
            { d: { color: color2 } },
            { _e: { color: color2 } },
            { _f_: { color: color2 } }, // eslint-disable-next-line camelcase
            { _g_1: { color: color2 } },
          ],
        });
      },
      autoSymbolsName: () => {
        eqn = new Equation(diagram.shapes, { color: color1, scale: 0.95 });
        eqn.addForms({
          0: { frac: ['a', 'vinculum', 'b'] },
        });
      },
      autoSymbolsID: () => {
        eqn = new Equation(diagram.shapes, { color: color1, scale: 0.95 });
        eqn.addForms({
          0: { frac: ['a', 'v_vinculum', 'b'] },
          // 2: ['a', 'b', { vinculum }],
        });
      },
      autoSymbolsObjectName: () => {
        eqn = new Equation(diagram.shapes, { color: color1, scale: 0.95 });
        eqn.addForms({
          0: { frac: ['a', { vinculum: { color: color2 } }, 'b'] },
        });
      },
      autoSymbolsObjectID: () => {
        eqn = new Equation(diagram.shapes, { color: color1, scale: 0.95 });
        eqn.addForms({  // eslint-disable-next-line camelcase
          0: { frac: ['a', { v_vinculum: { color: color2 } }, 'b'] },
        });
      },
    };
  });
  test('All Text in constructor', () => {
    ways.allTextInConstructor();
    expect(eqn).toHaveProperty('_a');
    expect(eqn).toHaveProperty('_b');
    expect(eqn).toHaveProperty('_c');
    expect(eqn).toHaveProperty('__2');
    expect(eqn).toHaveProperty('_v');

    // Check color
    expect(eqn._a.drawingObject.text[0].font.color)
      .toEqual(color1);

    // Check math vs number style
    expect(eqn._a.drawingObject.text[0].font.style).toBe('italic');
    expect(eqn.__2.drawingObject.text[0].font.style).toBe('normal');
  });
  test('All Text in constructor with all options', () => {
    ways.allTextInConstructorAllOptions();

    expect(eqn).toHaveProperty('_a');
    expect(eqn).toHaveProperty('_b');
    expect(eqn).toHaveProperty('_c');
    expect(eqn).toHaveProperty('__2');
    expect(eqn).toHaveProperty('_v');

    // Check color
    expect(eqn._a.drawingObject.text[0].font.color)
      .toEqual(color1);

    // Check position
    expect(eqn.transform.t()).toEqual(new Point(1, 1));

    // Check default form alignment
    expect(eqn.eqn.formDefaults.alignment.fixTo).toEqual(new Point(2, 2));
    expect(eqn.eqn.formDefaults.alignment.xAlign).toEqual('right');
    expect(eqn.eqn.formDefaults.alignment.yAlign).toEqual('top');
    expect(eqn.eqn.scale).toEqual(0.45);
    tools.cleanUIDs(eqn);
    eqn._a.parent = null;
    // console.log(eqn._a)
    expect(eqn._a).toMatchSnapshot();
  });
  test('Separate All Text', () => {
    ways.separateAllText();
    expect(eqn).toHaveProperty('_a');
    expect(eqn).toHaveProperty('_b');
    expect(eqn).toHaveProperty('_c');
    expect(eqn).toHaveProperty('__2');
    expect(eqn).toHaveProperty('_v');

    // Check color
    expect(eqn._a.drawingObject.text[0].font.color)
      .toEqual(color1);

    // Test locations of all elements
    tools.cleanUIDs(eqn);
    expect(round(eqn._a.transform.mat)).toMatchSnapshot();
    expect(round(eqn.__2.transform.mat)).toMatchSnapshot();
    expect(round(eqn._v.transform.mat)).toMatchSnapshot();
    expect(round(eqn._c.transform.mat)).toMatchSnapshot();
  });
  test('Equation Scale', () => {
    ways.equationScale();
    expect(eqn.eqn.scale).toBe(0.95);
    const h1 = eqn.eqn.forms['0'].content[0].height;
    const h2 = eqn.eqn.forms['1'].content[0].height;
    expect(round(h1 / h2)).toBe(round(0.95 / 0.85));
  });
  test('Equation Form Alignment', () => {
    ways.equationFormAlignment();
    eqn.showForm('originBasline');
    const originBasline = eqn._a.getPosition('diagram');
    eqn.showForm('aCenterTop');
    const aCenterTop = eqn._a.getPosition('diagram');
    eqn.showForm('aRightMiddle');
    const aRightMiddle = eqn._a.getPosition('diagram');
    eqn.showForm('bRightBottom');
    const bRightBottom = eqn._a.getPosition('diagram');

    // Test x positions
    expect(originBasline.x > aCenterTop.x).toBe(true);
    expect(aCenterTop.x > aRightMiddle.x).toBe(true);
    expect(aRightMiddle.x > bRightBottom.x).toBe(true);

    // Test y positions
    expect(aCenterTop.y < aRightMiddle.y).toBe(true);
    expect(aRightMiddle.y < originBasline.y).toBe(true);
    expect(originBasline.y < bRightBottom.y).toBe(true);
  });
  test('Phrases', () => {
    ways.phrases();
    expect(clean('0a')).toEqual(clean('0b'));
    //
    expect(clean('f')).toEqual(clean('f0'));
    expect(clean('f')).toEqual(clean('f1'));
    expect(clean('f')).toEqual(clean('f2'));
    expect(clean('f')).toEqual(clean('f3'));
    //
    expect(clean('d')).toEqual(clean('d0'));
    expect(clean('e')).toEqual(clean('e0'));
  });
  test('Auto Add Elements', () => {
    ways.autoAddElements();
    expect(eqn._a).not.toBe(undefined);
    expect(eqn._a.drawingObject.text[0].text).toBe('a');
    expect(eqn._a.color).toEqual(color1);

    expect(eqn._b_1).not.toBe(undefined);
    expect(eqn._b_1.drawingObject.text[0].text).toBe('b');
    expect(eqn._b_1.color).toEqual(color1);

    expect(eqn.__c_2).not.toBe(undefined);
    expect(eqn.__c_2.drawingObject.text[0].text).toBe('c');
    expect(eqn.__c_2.color).toEqual(color1);

    expect(eqn._d).not.toBe(undefined);
    expect(eqn._d.drawingObject.text[0].text).toBe('d');
    expect(eqn._d.color).toEqual(color2);

    expect(eqn.__e).not.toBe(undefined);
    expect(eqn.__e.drawingObject.text[0].text).toBe('e');
    expect(eqn.__e.color).toEqual(color2);

    expect(eqn.__f_).not.toBe(undefined);
    expect(eqn.__f_.drawingObject.text[0].text).toBe('f');
    expect(eqn.__f_.color).toEqual(color2);

    expect(eqn.__g_1).not.toBe(undefined);
    expect(eqn.__g_1.drawingObject.text[0].text).toBe('g');
    expect(eqn.__g_1.color).toEqual(color2);
  });
  test('Auto Symbols Name', () => {
    ways.autoSymbolsName();
    expect(eqn._a).not.toBe(undefined);
    expect(eqn._b).not.toBe(undefined);
    expect(eqn._a.drawingObject.text[0].text).toBe('a');
    expect(eqn._b.drawingObject.text[0].text).toBe('b');
    expect(eqn._vinculum.drawingObject.points).toEqual([
      0, 0, 0, 0.01, 0.133, 0, 0.133, 0.01,
    ]);
    expect(eqn._vinculum.color).toEqual(color1);
  });
  test('Auto Symbols ID', () => {
    ways.autoSymbolsID();
    expect(eqn._v.drawingObject.points).toEqual([
      0, 0, 0, 0.01, 0.133, 0, 0.133, 0.01,
    ]);
    expect(eqn._v.color).toEqual(color1);
  });
  test('Auto Symbols Object Name', () => {
    ways.autoSymbolsObjectName();
    expect(eqn._vinculum.drawingObject.points).toEqual([
      0, 0, 0, 0.01, 0.133, 0, 0.133, 0.01,
    ]);
    expect(eqn._vinculum.color).toEqual(color1);
  });
  test('Auto Symbols Object ID', () => {
    ways.autoSymbolsObjectID();
    expect(eqn._v.drawingObject.points).toEqual([
      0, 0, 0, 0.01, 0.133, 0, 0.133, 0.01,
    ]);
    expect(eqn._v.color).toEqual(color1);
  });
});
