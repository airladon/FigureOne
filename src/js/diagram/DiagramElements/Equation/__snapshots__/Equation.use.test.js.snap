// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Different ways to make an equation All Text in constructor with all options 1`] = `
DiagramElementPrimative {
  "angleToDraw": -1,
  "animate": Object {
    "color": Object {
      "callback": null,
      "plan": Array [],
      "toDisolve": "",
    },
    "custom": Object {
      "callback": null,
      "plan": Array [],
    },
    "transform": Object {
      "callback": null,
      "plan": Array [],
      "translation": Object {
        "options": Object {
          "controlPoint": null,
          "direction": "",
          "magnitude": 0.5,
          "offset": 0.5,
          "rot": 1,
        },
        "style": "linear",
      },
    },
  },
  "cannotTouchHole": false,
  "color": Array [
    0.9490196078431372,
    0,
    0,
    1,
  ],
  "diagramLimits": Rect {
    "bottom": -1,
    "height": 2,
    "left": -1,
    "right": 1,
    "top": 1,
    "width": 2,
  },
  "drawPriority": 1,
  "drawingObject": TextObject {
    "border": Array [
      Array [
        Point {
          "x": 0,
          "y": 0.038000000000000006,
        },
        Point {
          "x": 0.04000000000000001,
          "y": 0.038000000000000006,
        },
        Point {
          "x": 0.04000000000000001,
          "y": -0.0032000000000000006,
        },
        Point {
          "x": 0,
          "y": -0.0032000000000000006,
        },
      ],
    ],
    "drawContext2D": DrawContext2D {
      "canvas": Object {
        "height": 1000,
        "offsetHeight": 500,
        "offsetWidth": 1000,
        "style": Object {
          "height": "500px",
          "width": "1000px",
        },
        "width": 2000,
      },
      "ctx": Object {
        "canvas": Object {
          "height": 500,
          "width": 1000,
        },
        "clearRect": [Function],
        "fillText": [Function],
        "filledText": Object {
          "count": 0,
          "text": "",
          "x": null,
          "y": null,
        },
        "font": "italic 200 100px Times New Roman",
        "measureText": [Function],
        "restore": [Function],
        "save": [Function],
        "scale": [Function],
        "textAlign": "left",
        "textBaseline": "alphabetic",
        "transform": [Function],
        "transformMatrix": Array [
          0,
        ],
      },
      "ratio": 2,
    },
    "holeBorder": Array [
      Array [],
    ],
    "location": Point {
      "x": 0,
      "y": 0,
    },
    "scalingFactor": 499.99999999999994,
    "text": Array [
      DiagramText {
        "font": DiagramFont {
          "alignH": "left",
          "alignV": "alphabetic",
          "color": "rgba(242,0,0,1)",
          "family": "Times New Roman",
          "size": 0.2,
          "style": "italic",
          "weight": "200",
        },
        "location": Point {
          "x": 0,
          "y": 0,
        },
        "text": "a",
      },
    ],
  },
  "hasTouchableElements": false,
  "interactiveLocation": Point {
    "x": 0,
    "y": 0,
  },
  "isInteractive": false,
  "isMovable": false,
  "isShown": true,
  "isTouchable": false,
  "lastDrawElementTransformPosition": Object {
    "elementCount": 2,
    "parentCount": 6,
  },
  "lastDrawTransform": Transform {
    "index": 0,
    "mat": Array [
      0.45,
      0,
      2.8825849999999997,
      0,
      0.45,
      2.890505,
      0,
      0,
      1,
    ],
    "name": "Equation",
    "order": Array [
      Translation {
        "name": "",
        "x": 1.9612999999999998,
        "y": 1.9788999999999999,
      },
      Scale {
        "name": "",
        "x": 0.45,
        "y": 0.45,
      },
      Scale {
        "name": "Equation",
        "x": 1,
        "y": 1,
      },
      Rotation {
        "name": "Equation",
        "r": 0,
      },
      Translation {
        "name": "Equation",
        "x": 1,
        "y": 1,
      },
      Scale {
        "name": "Equation",
        "x": 1,
        "y": 1,
      },
      Rotation {
        "name": "Equation",
        "r": 0,
      },
      Translation {
        "name": "Equation",
        "x": 1,
        "y": 1,
      },
    ],
  },
  "lengthToDraw": -1,
  "move": Object {
    "bounce": true,
    "canBeMovedAfterLoosingTouch": false,
    "element": null,
    "freely": Object {
      "deceleration": TransformLimit {
        "rotation": 5,
        "scale": 5,
        "translation": 5,
      },
      "zeroVelocityThreshold": TransformLimit {
        "rotation": 0.001,
        "scale": 0.001,
        "translation": 0.001,
      },
    },
    "limitLine": null,
    "limitToDiagram": false,
    "maxTransform": Transform {
      "index": 2,
      "mat": Array [
        1000,
        0,
        1000,
        0,
        1000,
        1000,
        0,
        0,
        1,
      ],
      "name": "",
      "order": Array [
        Scale {
          "name": "",
          "x": 1000,
          "y": 1000,
        },
        Translation {
          "name": "",
          "x": 1000,
          "y": 1000,
        },
      ],
    },
    "maxVelocity": TransformLimit {
      "rotation": 5,
      "scale": 5,
      "translation": 5,
    },
    "minTransform": Transform {
      "index": 2,
      "mat": Array [
        -1000,
        0,
        -1000,
        0,
        -1000,
        -1000,
        0,
        0,
        1,
      ],
      "name": "",
      "order": Array [
        Scale {
          "name": "",
          "x": -1000,
          "y": -1000,
        },
        Translation {
          "name": "",
          "x": -1000,
          "y": -1000,
        },
      ],
    },
    "type": "translation",
  },
  "name": "a",
  "noRotationFromParent": false,
  "onClick": null,
  "pointsToDraw": -1,
  "pulse": Object {
    "A": 1,
    "B": 0.5,
    "C": 0,
    "callback": [Function],
    "frequency": 0.5,
    "num": 1,
    "style": [Function],
    "time": 1,
    "transformMethod": [Function],
  },
  "setTransformCallback": [Function],
  "state": Object {
    "animation": Object {
      "currentPhase": AnimationPhase {
        "animationStyle": [Function],
        "callback": null,
        "deltaTransform": Transform {
          "index": 0,
          "mat": Array [
            1,
            0,
            0,
            0,
            1,
            0,
            0,
            0,
            1,
          ],
          "name": "",
          "order": Array [],
        },
        "finishOnCancel": true,
        "rotDirection": 0,
        "startTime": -1,
        "startTransform": null,
        "targetTransform": Transform {
          "index": 0,
          "mat": Array [
            1,
            0,
            0,
            0,
            1,
            0,
            0,
            0,
            1,
          ],
          "name": "",
          "order": Array [],
        },
        "timeOrVelocity": 1,
        "translationOptions": Object {
          "controlPoint": null,
          "direction": "",
          "magnitude": 0.5,
          "offset": 0.5,
          "rot": 1,
        },
        "translationStyle": "linear",
      },
      "currentPhaseIndex": 0,
    },
    "colorAnimation": Object {
      "currentPhase": ColorAnimationPhase {
        "animationStyle": [Function],
        "callback": null,
        "deltaColor": Array [
          0,
          0,
          0,
          1,
        ],
        "disolve": null,
        "endColor": Array [
          0,
          0,
          0,
          1,
        ],
        "finishOnCancel": true,
        "startColor": null,
        "startTime": -1,
        "targetColor": Array [
          0,
          0,
          0,
          1,
        ],
        "time": 1,
      },
      "currentPhaseIndex": 0,
    },
    "customAnimation": Object {
      "currentPhase": CustomAnimationPhase {
        "animationCallback": [Function],
        "animationStyle": [Function],
        "callback": null,
        "finishOnCancel": true,
        "plannedStartTime": -0,
        "startPercent": 0,
        "startTime": -1,
        "time": 1,
      },
      "currentPhaseIndex": 0,
    },
    "disolving": "",
    "isAnimating": false,
    "isAnimatingColor": false,
    "isAnimatingCustom": false,
    "isBeingMoved": false,
    "isMovingFreely": false,
    "isPulsing": false,
    "movement": Object {
      "previousTime": -1,
      "previousTransform": Transform {
        "index": 2,
        "mat": Array [
          1,
          0,
          0,
          0,
          1,
          0,
          0,
          0,
          1,
        ],
        "name": "",
        "order": Array [
          Scale {
            "name": "",
            "x": 1,
            "y": 1,
          },
          Translation {
            "name": "",
            "x": 0,
            "y": 0,
          },
        ],
      },
      "velocity": Transform {
        "index": 2,
        "mat": Array [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          1,
        ],
        "name": "",
        "order": Array [
          Scale {
            "name": "",
            "x": 0,
            "y": 0,
          },
          Translation {
            "name": "",
            "x": 0,
            "y": 0,
          },
        ],
      },
    },
    "pulse": Object {
      "startTime": -1,
    },
  },
  "transform": Transform {
    "index": 2,
    "mat": Array [
      0.45,
      0,
      1.9612999999999998,
      0,
      0.45,
      1.9788999999999999,
      0,
      0,
      1,
    ],
    "name": "",
    "order": Array [
      Scale {
        "name": "",
        "x": 0.45,
        "y": 0.45,
      },
      Translation {
        "name": "",
        "x": 1.9612999999999998,
        "y": 1.9788999999999999,
      },
    ],
  },
  "uid": "0000000000",
}
`;

exports[`Different ways to make an equation Separate All Text 1`] = `
Array [
  0.7,
  0,
  0.0042,
  0,
  0.7,
  0.08624,
  0,
  0,
  1,
]
`;

exports[`Different ways to make an equation Separate All Text 2`] = `
Array [
  0.7,
  0,
  0.0042,
  0,
  0.7,
  -0.0252,
  0,
  0,
  1,
]
`;

exports[`Different ways to make an equation Separate All Text 3`] = `
Array [
  0.0364,
  0,
  0,
  0,
  0.008,
  0.049,
  0,
  0,
  1,
]
`;

exports[`Different ways to make an equation Separate All Text 4`] = `
Array [
  0.7,
  0,
  0.0364,
  0,
  0.7,
  0,
  0,
  0,
  1,
]
`;
